Streamline size estimates

nnethercote opened the pull request:
Makes things nicer and a tiny bit faster.

r? @wesleywiser

nnethercote committed:
Store item size estimate in `MonoItemData`.

This means we call `MonoItem::size_estimate` (which involves a query)
less often: just once per mono item, and then once more per inline item
placement. After that we can reuse the stored value as necessary. This
means `CodegenUnit::compute_size_estimate` is cheaper.

nnethercote committed:
Introduce `MonoItemData`.

It replaces `(Linkage, Visibility)`, making the code nicer. Plus the
next commit will add another field.

nnethercote committed:
Remove `instance_def_size_estimate` query.

It doesn't seem worthwhile now that `MonoItem::size_estimate` is called
much less often.

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit adcfd93f0aa9b387f91a10e9dc49c5dc189572a0 with merge eb0596b60545ce2c3ac92dd2aafb5713910aeca4...
<!-- homu: {"type":"TryBuildStarted","head_sha":"adcfd93f0aa9b387f91a10e9dc49c5dc189572a0","merge_sha":"eb0596b60545ce2c3ac92dd2aafb5713910aeca4"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5552935892/jobs/10142232404)
Build commit: eb0596b60545ce2c3ac92dd2aafb5713910aeca4 (`eb0596b60545ce2c3ac92dd2aafb5713910aeca4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5552935892/jobs/10142232404"},"merge_sha":"eb0596b60545ce2c3ac92dd2aafb5713910aeca4"} -->

rust-timer commented:
Queued eb0596b60545ce2c3ac92dd2aafb5713910aeca4 with parent 320b412f9c55bf480d26276ff0ab480e4ecb29c0, future [comparison URL](https://perf.rust-lang.org/compare.html?start=320b412f9c55bf480d26276ff0ab480e4ecb29c0&end=eb0596b60545ce2c3ac92dd2aafb5713910aeca4).
<!-- rust-timer: temporary -->

nnethercote committed:
Ignore unreachable inlined items in `debug_dump`.

They're quite rare, and ignoring them simplifies things quite a bit, and
further reduces the number of calls to `MonoItem::size_estimate` to the
number of placed items (one per root item, and one or more per reachable
inlined item).

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit f2d28cc15b1a20c5a5b93031eb5d3cf110c61497 with merge a861e61c1dc3fcf632860ee8946626d8945d9bd9...
<!-- homu: {"type":"TryBuildStarted","head_sha":"f2d28cc15b1a20c5a5b93031eb5d3cf110c61497","merge_sha":"a861e61c1dc3fcf632860ee8946626d8945d9bd9"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5553903030/jobs/10144555082)
Build commit: a861e61c1dc3fcf632860ee8946626d8945d9bd9 (`a861e61c1dc3fcf632860ee8946626d8945d9bd9`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5553903030/jobs/10144555082"},"merge_sha":"a861e61c1dc3fcf632860ee8946626d8945d9bd9"} -->

wesleywiser commented on compiler/rustc_middle/src/mir/mono.rs:
```diff
@@ -230,14 +230,21 @@ pub struct CodegenUnit<'tcx> {
 
+/// Auxiliary info about a `MonoItem`.
+#[derive(Copy, Clone, PartialEq, Debug, HashStable)]
+pub struct MonoItemAux {
```
Nit: I think we would usually name this `MonoItemData`. I don't feel strongly about the current name.

nnethercote replied:
I have renamed it.

wesleywiser commented on compiler/rustc_middle/src/mir/mono.rs:
```diff
@@ -59,12 +59,19 @@ impl<'tcx> MonoItem<'tcx> {
-                // it contains.
-                tcx.instance_def_size_estimate(instance.def)
+                match instance.def {
+                    // "Normal" functions size estimate: the number of
+                    // statements, plus one for the terminator.
```
Oh, seeing this code again reminded me of something that occurred to me while reading your recent blog post: should we have different size estimates for codegen with and without optimizations enabled? 

I could believe this current estimate is relatively accurate for debug builds but many optimization passes in LLVM are super linear with regards to the number of basic blocks. Perhaps it makes sense to use different estimation strategies when we are optimizing code and not optimizing code?

nnethercote replied:
I have tried a variety of alternative estimation functions. #113047 and #113649 are two that I filed PRs for, and I've tried a bunch more locally. I can't say that I ever really saw meaningful differences between release and debug builds. Cleverer estimation functions have gotten me exactly zero wins at the moment, so I'm inclined to not spend more effort there :)

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit f4c46025646536123f570001cc322549a1623d7f with merge bef6ff618f17398775e9d8cd23a6f47d983869dc...
<!-- homu: {"type":"TryBuildStarted","head_sha":"f4c46025646536123f570001cc322549a1623d7f","merge_sha":"bef6ff618f17398775e9d8cd23a6f47d983869dc"} -->

rust-timer commented:
Finished benchmarking commit (eb0596b60545ce2c3ac92dd2aafb5713910aeca4): [comparison URL](https://perf.rust-lang.org/compare.html?start=320b412f9c55bf480d26276ff0ab480e4ecb29c0&end=eb0596b60545ce2c3ac92dd2aafb5713910aeca4&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=320b412f9c55bf480d26276ff0ab480e4ecb29c0&end=eb0596b60545ce2c3ac92dd2aafb5713910aeca4&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=320b412f9c55bf480d26276ff0ab480e4ecb29c0&end=eb0596b60545ce2c3ac92dd2aafb5713910aeca4&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 1.3% | [1.3%, 1.3%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=320b412f9c55bf480d26276ff0ab480e4ecb29c0&end=eb0596b60545ce2c3ac92dd2aafb5713910aeca4&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.7% | [-2.7%, -2.7%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=320b412f9c55bf480d26276ff0ab480e4ecb29c0&end=eb0596b60545ce2c3ac92dd2aafb5713910aeca4&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 657.803s -> 657.535s (-0.04%)

nnethercote commented:
@bors r=wesleywiser

bors commented:
:pushpin: Commit f4c46025646536123f570001cc322549a1623d7f has been approved by `wesleywiser`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=wesleywiser f4c46025646536123f570001cc322549a1623d7f -->
<!-- homu: {"type":"Approved","sha":"f4c46025646536123f570001cc322549a1623d7f","approver":"wesleywiser","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5559202105/jobs/10155771699)
Approved by: wesleywiser
Pushing bef6ff618f17398775e9d8cd23a6f47d983869dc to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"wesleywiser","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5559202105/jobs/10155771699"},"merge_sha":"bef6ff618f17398775e9d8cd23a6f47d983869dc"} -->

rust-timer commented:
Finished benchmarking commit (bef6ff618f17398775e9d8cd23a6f47d983869dc): [comparison URL](https://perf.rust-lang.org/compare.html?start=ad963232d9b987d66a6f8e6ec4141f672b8b9900&end=bef6ff618f17398775e9d8cd23a6f47d983869dc&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=ad963232d9b987d66a6f8e6ec4141f672b8b9900&end=bef6ff618f17398775e9d8cd23a6f47d983869dc&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ad963232d9b987d66a6f8e6ec4141f672b8b9900&end=bef6ff618f17398775e9d8cd23a6f47d983869dc&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.6%  | [2.6%, 2.6%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.8% | [-6.1%, -1.8%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ad963232d9b987d66a6f8e6ec4141f672b8b9900&end=bef6ff618f17398775e9d8cd23a6f47d983869dc&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=ad963232d9b987d66a6f8e6ec4141f672b8b9900&end=bef6ff618f17398775e9d8cd23a6f47d983869dc&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 657.64s -> 657.713s (0.01%)

the8472 commented:
There has been a bors issue that lead to the merge commit of this PR getting purged from master.
You'll have to make a new PR to reapply it.
